File: Data Structures & Algorithms/longest-common-subsequence/submission-0.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        t1 = len(text1)
        t2 = len(text2)

        memo = {}

        def dfs(i: int, j: int):
            # end of word(s)
            if i == t1 or j == t2:
                return 0

            if (i, j) in memo:
                return memo[(i, j)]

            if text1[i] == text2[j]:
                memo[(i, j)] = 1 + dfs(i + 1, j + 1)
            else:
                memo[(i, j)] = max(dfs(i + 1, j), dfs(i, j + 1))

            return memo[(i, j)]

        return dfs(
            i=0,
            j=0,
        )

File: Data Structures & Algorithms/redistribute-characters-to-make-all-strings-equal/submission-0.cpp
class Solution {
public:
    bool makeEqual(vector<string>& words) {
        vector<int> freq(26, 0);
        int n = words.size();

        for(const string& word: words)
            for(char c : word)
                freq[c-'a']++;
        
        for(int f : freq) 
            if(f % n != 0) 
                return false;

        return true;
    }
};